### PR TITLE
Force to recaptcha verification flow for phone auth in Simulators

### PR DESCRIPTION
**Pull request by renkelvin**
In the latest Xcode and macOS versions, APNs is now supported in simulators. Before that, phone auth will always fallback to the recaptcha verification flow.
Since FCM doesn't support iOS 16 simulators yet, https://github.com/firebase/firebase-ios-sdk/issues/9968, this change force the phone auth flow fallback to recaptcha in simulators.

### CODE REVIEW

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 1]
```diff
@@ -593,6 +595,10 @@ - (void)verifyClientAndSendVerificationCodeToPhoneNumber:(NSString *)phoneNumber
  */
 - (void)verifyClientWithUIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
                         completion:(FIRVerifyClientCallback)completion {
+  if ([GULAppEnvironmentUtil isSimulator]) {
```
Comment (paulb777) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:598]: Should there be an iOS version check for at least 16 as well?

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 2]
```diff
@@ -593,6 +595,10 @@ - (void)verifyClientAndSendVerificationCodeToPhoneNumber:(NSString *)phoneNumber
...
+    [self reCAPTCHAFlowWithUIDelegate:UIDelegate completion:completion];
```
Comment (rosalyntan) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:599]: I'm wondering if there's actually something going wrong in our fallback logic:

As per the reported issue, APNS is failing with INVALID_APP_CREDENTIAL: https://github.com/firebase/firebase-ios-sdk/issues/10214#issue-1373420565. However, we should be falling back to reCAPTCHA automatically in that case: https://github.com/firebase/firebase-ios-sdk/blob/bba19ad013bdd08eae551636e9e345642afb24a9/FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m#L624

That might be worth looking into a little before hard-coding simulators to always use reCAPTCHA.

Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:599]: I don't think we want to fallback to recaptcha in that case. If APNs is available but the app credential is invalid, there is a good chance the the app is not legitimate, so we should just error out.

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m (hunk 1 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m]: Yeah, good point.

Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m]: Added a check to not include this change in unit tests (as they're run in the simulators). Please take a look. Thanks!

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 3]
```diff
@@ -593,6 +595,16 @@ - (void)verifyClientAndSendVerificationCodeToPhoneNumber:(NSString *)phoneNumber
  */
 - (void)verifyClientWithUIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
                         completion:(FIRVerifyClientCallback)completion {
+  // Remove the simulator check below after FCM supports APNs in simulators
```
Comment (paulb777) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:598]: Why does Auth depend on FCM?

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 4]
```diff
@@ -593,6 +595,16 @@ - (void)verifyClientAndSendVerificationCodeToPhoneNumber:(NSString *)phoneNumber
...
+  if (@available(iOS 17, *)) {
```
Comment (paulb777) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:599]: How will this change make any difference for iOS 16 simulator users?

**Review by paulb777 — COMMENTED** [review]
It looks like the integration test Firebase project is in a bad state for the signin tests. Would you fix?

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m (hunk 3 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:598]: Auth uses FCM to send a silent push notification which contains a secret for verification. It's not delivered in the simulators for now.

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m (hunk 4 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m]: Ops, wrong commit. Just fixed.

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 5]
```diff
@@ -593,6 +595,16 @@ - (void)verifyClientAndSendVerificationCodeToPhoneNumber:(NSString *)phoneNumber
  */
 - (void)verifyClientWithUIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
                         completion:(FIRVerifyClientCallback)completion {
+  // Remove the simulator check below after FCM supports APNs in simulators
+  if (@available(iOS 16, *)) {
+    if ([GULAppEnvironmentUtil isSimulator]) {
```
Comment (morganchen12) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:600]: You can use `#if TARGET_OS_SIMULATOR` so this bit of code doesn't get compiled into non-simulator target destination builds.

**Review by morganchen12 — APPROVED** [review]

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m [hunk 6]
```diff
@@ -14,7 +14,9 @@
  * limitations under the License.
  */
 
+#import <GoogleUtilities/GULAppEnvironmentUtil.h>
```
Comment (paulb777) [inline FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m:17]: This line may not be needed any longer.

**Review by paulb777 — APPROVED** [review]

File: FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m (hunk 5 above)
Comment (renkelvin) [reply FirebaseAuth/Sources/AuthProvider/Phone/FIRPhoneAuthProvider.m]: Yeah, good point

**Comment by renkelvin** [comment]
> It looks like the integration test Firebase project is in a bad state for the signin tests. Would you fix?

Yeah, it's fixed now.

**Pull request merged**